File: vision_core/src/camera_manager.py
"""
XNav Camera Manager
Manages camera capture, settings, and frame delivery.
Supports V4L2 cameras on Raspberry Pi CM.
"""

import cv2
import threading
import time
import logging
import numpy as np
from typing import Optional, Tuple, Callable

logger = logging.getLogger(__name__)


class CameraManager:
    """Manages a camera capture device and exposes frames to consumers."""

    def __init__(self, config_manager):
        self._cfg = config_manager
        self._cap: Optional[cv2.VideoCapture] = None
        self._lock = threading.Lock()
        self._latest_frame: Optional[np.ndarray] = None
        self._latest_gray: Optional[np.ndarray] = None
        self._frame_time: float = 0.0
        self._running = False
        self._thread: Optional[threading.Thread] = None
        self._fps_actual: float = 0.0
        self._frame_count: int = 0
        self._fps_t0: float = 0.0
        self._on_frame_callbacks: list = []

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    def start(self):
        self._running = True
        self._thread = threading.Thread(target=self._capture_loop, daemon=True, name="CamCapture")
        self._thread.start()
        logger.info("Camera manager started")

    def stop(self):
        self._running = False
        if self._thread:
            self._thread.join(timeout=3)
        if self._cap:
            self._cap.release()
        logger.info("Camera manager stopped")

    def restart(self):
        self.stop()
        time.sleep(0.5)
        self.start()

    # ------------------------------------------------------------------
    # Frame access
    # ------------------------------------------------------------------

    def get_frame(self) -> Tuple[Optional[np.ndarray], Optional[np.ndarray], float]:
        """Returns (color_frame, gray_frame, timestamp)."""
        with self._lock:
            return (
                self._latest_frame.copy() if self._latest_frame is not None else None,
                self._latest_gray.copy() if self._latest_gray is not None else None,
                self._frame_time
            )

    def get_fps(self) -> float:
        return self._fps_actual

    def register_frame_callback(self, cb: Callable):
        """Register callback(color, gray, timestamp) for each new frame."""
        self._on_frame_callbacks.append(cb)

    # ------------------------------------------------------------------
    # Settings
    # ------------------------------------------------------------------

    def apply_settings(self):
        """Apply current config camera settings to the capture device."""
        if self._cap is None or not self._cap.isOpened():
            return
        cam = self._cfg.get("camera")
        if cam is None:
            return

        # Exposure (disable auto first)
        auto_exp = cam.get("auto_exposure", False)
        # V4L2 auto exposure: 1 = manual, 3 = auto
        self._cap.set(cv2.CAP_PROP_AUTO_EXPOSURE, 3 if auto_exp else 1)
        if not auto_exp:
            self._cap.set(cv2.CAP_PROP_EXPOSURE, cam.get("exposure", 100))

        self._cap.set(cv2.CAP_PROP_GAIN, cam.get("gain", 50))
        self._cap.set(cv2.CAP_PROP_BRIGHTNESS, cam.get("brightness", 50))
        self._cap.set(cv2.CAP_PROP_CONTRAST, cam.get("contrast", 50))
        logger.info("Camera settings applied")

    def get_jpeg_frame(self, quality: int = 70) -> Optional[bytes]:
        """Return latest frame as JPEG bytes for MJPEG streaming."""
        with self._lock:
            frame = self._latest_frame
        if frame is None:
            return None
        ret, buf = cv2.imencode(".jpg", frame, [cv2.IMWRITE_JPEG_QUALITY, quality])
        return buf.tobytes() if ret else None

    # ------------------------------------------------------------------
    # Internal
    # ------------------------------------------------------------------

    def _open_camera(self):
        cam = self._cfg.get("camera")
        device = cam.get("device", "/dev/video0")
        width = cam.get("width", 1280)
        height = cam.get("height", 720)
        fps = cam.get("fps", 90)
        idx = cam.get("camera_index", 0)

        # Try device path first, fallback to index
        for src in [device, idx]:
            cap = cv2.VideoCapture(src, cv2.CAP_V4L2)
            if cap.isOpened():
                cap.set(cv2.CAP_PROP_FRAME_WIDTH, width)
                cap.set(cv2.CAP_PROP_FRAME_HEIGHT, height)
                cap.set(cv2.CAP_PROP_FPS, fps)
                # Use MJPG codec for higher FPS
                cap.set(cv2.CAP_PROP_FOURCC, cv2.VideoWriter_fourcc(*"MJPG"))
                # Minimize buffer for low latency
                cap.set(cv2.CAP_PROP_BUFFERSIZE, 1)
                logger.info("Camera opened: src=%s res=%dx%d fps=%d", src, width, height, fps)
                return cap
            cap.release()

        logger.error("Failed to open camera")
        return None

    def _capture_loop(self):
        self._cap = self._open_camera()
        if self._cap:
            self.apply_settings()
        self._fps_t0 = time.monotonic()
        self._frame_count = 0

        while self._running:
            if self._cap is None or not self._cap.isOpened():
                logger.warning("Camera not open, retrying in 2s...")
                time.sleep(2)
                self._cap = self._open_camera()
                if self._cap:
                    self.apply_settings()
                continue

            ret, frame = self._cap.read()
            if not ret:
                logger.warning("Failed to capture frame, retrying...")
                time.sleep(0.05)
                continue

            ts = time.monotonic()
            gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)

            with self._lock:
                self._latest_frame = frame
                self._latest_gray = gray
                self._frame_time = ts

            # FPS calculation
            self._frame_count += 1
            elapsed = ts - self._fps_t0
            if elapsed >= 1.0:
                self._fps_actual = self._frame_count / elapsed
                self._frame_count = 0
                self._fps_t0 = ts

            # Fire callbacks
            for cb in self._on_frame_callbacks:
                try:
                    cb(frame, gray, ts)
                except Exception as e:
                    logger.warning("Frame callback error: %s", e)

        if self._cap:
            self._cap.release()
            self._cap = None

File: vision_core_cpp/src/camera_manager.cpp
#include "camera_manager.hpp"

#include <syslog.h>
#include <chrono>
#include <thread>
#include <opencv2/imgproc.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/videoio.hpp>

// V4L2 auto-exposure control values (CAP_PROP_AUTO_EXPOSURE)
static constexpr double V4L2_EXPOSURE_MANUAL = 1.0;
static constexpr double V4L2_EXPOSURE_AUTO   = 3.0;

CameraManager::CameraManager(ConfigManager& cfg)
    : m_cfg(cfg)
{
    m_fpsT0 = std::chrono::steady_clock::now();
}

CameraManager::~CameraManager() {
    stop();
}

void CameraManager::start() {
    m_running = true;
    m_thread = std::thread(&CameraManager::captureLoop, this);
    syslog(LOG_INFO, "Camera manager started");
}

void CameraManager::stop() {
    m_running = false;
    if (m_thread.joinable()) {
        m_thread.join();
    }
    if (m_cap.isOpened()) {
        m_cap.release();
    }
    syslog(LOG_INFO, "Camera manager stopped");
}

void CameraManager::restart() {
    stop();
    std::this_thread::sleep_for(std::chrono::milliseconds(500));
    start();
}

void CameraManager::registerFrameCallback(FrameCallback cb) {
    m_callbacks.push_back(std::move(cb));
}

bool CameraManager::getFrame(cv::Mat& frame, cv::Mat& gray, double& timestamp) const {
    std::lock_guard<std::mutex> lock(m_frameLock);
    if (m_latestFrame.empty()) return false;
    frame = m_latestFrame.clone();
    gray = m_latestGray.clone();
    timestamp = m_frameTime;
    return true;
}

std::optional<std::vector<uint8_t>> CameraManager::getJpegFrame(int quality) const {
    std::lock_guard<std::mutex> lock(m_frameLock);
    if (m_latestFrame.empty()) return std::nullopt;
    cv::Mat frame = m_latestFrame.clone();

    std::vector<uint8_t> buf;
    std::vector<int> params = {cv::IMWRITE_JPEG_QUALITY, quality};
    if (cv::imencode(".jpg", frame, buf, params)) {
        return buf;
    }
    return std::nullopt;
}

void CameraManager::applySettings() {
    if (!m_cap.isOpened()) return;
    auto cam = m_cfg.getSection("camera");

    bool auto_exp = cam.value("auto_exposure", false);
    // V4L2: 1 = manual, 3 = auto
    m_cap.set(cv::CAP_PROP_AUTO_EXPOSURE, auto_exp ? V4L2_EXPOSURE_AUTO : V4L2_EXPOSURE_MANUAL);
    if (!auto_exp) {
        m_cap.set(cv::CAP_PROP_EXPOSURE, cam.value("exposure", 100));
    }
    m_cap.set(cv::CAP_PROP_GAIN,       cam.value("gain", 50));
    m_cap.set(cv::CAP_PROP_BRIGHTNESS, cam.value("brightness", 50));
    m_cap.set(cv::CAP_PROP_CONTRAST,   cam.value("contrast", 50));
    syslog(LOG_INFO, "Camera settings applied");
}

bool CameraManager::openCamera() {
    auto cam = m_cfg.getSection("camera");
    std::string device = cam.value("device", "/dev/video0");
    int width  = cam.value("width",  1280);
    int height = cam.value("height", 720);
    int fps    = cam.value("fps",    90);

    if (m_cap.isOpened()) m_cap.release();

    m_cap.open(device, cv::CAP_V4L2);
    if (!m_cap.isOpened()) {
        // Fallback: try by index 0
        m_cap.open(0, cv::CAP_V4L2);
    }
    if (!m_cap.isOpened()) {
        syslog(LOG_ERR, "Failed to open camera: %s", device.c_str());
        return false;
    }

    m_cap.set(cv::CAP_PROP_FOURCC, cv::VideoWriter::fourcc('M','J','P','G'));
    m_cap.set(cv::CAP_PROP_FRAME_WIDTH,  width);
    m_cap.set(cv::CAP_PROP_FRAME_HEIGHT, height);
    m_cap.set(cv::CAP_PROP_FPS,          fps);
    m_cap.set(cv::CAP_PROP_BUFFERSIZE,   1);

    applySettings();
    syslog(LOG_INFO, "Camera opened: %s %dx%d @%dfps", device.c_str(), width, height, fps);
    return true;
}

void CameraManager::captureLoop() {
    m_fpsT0 = std::chrono::steady_clock::now();
    m_frameCount = 0;

    while (m_running) {
        if (!m_cap.isOpened()) {
            if (!openCamera()) {
                std::this_thread::sleep_for(std::chrono::seconds(2));
                continue;
            }
        }

        cv::Mat frame;
        if (!m_cap.read(frame) || frame.empty()) {
            syslog(LOG_WARNING, "Camera read failed, retrying...");
            m_cap.release();
            std::this_thread::sleep_for(std::chrono::milliseconds(200));
            continue;
        }

        double ts = std::chrono::duration<double>(
            std::chrono::steady_clock::now().time_since_epoch()).count();

        cv::Mat gray;
        cv::cvtColor(frame, gray, cv::COLOR_BGR2GRAY);

        {
            std::lock_guard<std::mutex> lock(m_frameLock);
            m_latestFrame = frame;
            m_latestGray  = gray;
            m_frameTime   = ts;
        }

        // FPS calculation
        m_frameCount++;
        auto now = std::chrono::steady_clock::now();
        double elapsed = std::chrono::duration<double>(now - m_fpsT0).count();
        if (elapsed >= 1.0) {
            m_fps = static_cast<float>(m_frameCount / elapsed);
            m_frameCount = 0;
            m_fpsT0 = now;
        }

        // Fire callbacks
        for (auto& cb : m_callbacks) {
            cb(frame, gray, ts);
        }
    }
}
